plugins/home/src/extensions.tsx

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { lazy, Suspense } from 'react';
+import { IconButton } from '@material-ui/core';
+import SettingsIcon from '@material-ui/icons/Settings';
+import { InfoCard } from '@backstage/core-components';
+import { SettingsModal } from './components';
+import { createReactExtension, useApp } from '@backstage/core-plugin-api';
+
+const lazyLoadedComponent = (
+  component: () => Promise<(props: any) => JSX.Element>,
+) => lazy(() => component().then(inner => ({ default: inner })));
+
+export function createHomePageComponentExtension({
+  title,
+  component,
+  actions,
+  contextProvider,
+  settings,
+}: {
+  title: string;
+  component: () => Promise<(props: any) => JSX.Element>;
+  actions?: () => Promise<(props: any) => JSX.Element>;
+  contextProvider?: () => Promise<(props: any) => JSX.Element>;
+  settings?: () => Promise<(props: any) => JSX.Element>;
+}) {
+  const Content = lazyLoadedComponent(component);
+  const Actions = actions ? lazyLoadedComponent(actions) : () => null;
+  const Settings = settings ? lazyLoadedComponent(settings) : () => null;
+  const ContextProvider = contextProvider
+    ? lazyLoadedComponent(contextProvider)
+    : () => null;
+
+  const HomePageExtension = () => {
```

Comment:
@Rugvip I think this is where I'd like to get your input here, weather letting the Extension render itself is a good idea, or weather all this data should be passed as Extension data to the parent `HomePage` component wrapper and it does the rendering by collecting all the data it needs for each extension.

It think the problem with that is that we might have to do some magic to get it to work with the other components that the user provides like the grid layout etc.

Also thought's on the interface being split apart so each part of the card is `lazy`? I thought that sometimes maybe you don't want to show actions for a card, so why should be loading in the octokit library upfront or whatever wheb you never use it.

It's a little more overhead on the end user, but maybe it's ok?

Reply:
We've been given a lot of consideration to the API of the App Integrator. The way we have it now, the HomePage Components will be rendered like
```
<RandomJokeComponent />
<RecentPRsComponent />
```

I think there's a lot to be said for breaking out the HomePageComponent JSX we currently have hidden within the extension, so that the API instead becomes
```
<HomePageComponent component={RandomJoke} />
<HomePageComponent component={RecentPRs} />
```

The primary advantages to this are:

1. It provides a clearer interface to the integrator, it is more obvious that these two components will render the same way, and perhaps take some props that are the same such as `theme`, `rowSpan`, etc.
2. It makes it so that the App Integrator or anyone else for that matter can develop their own HomePage Component renderer, such as an Accordion, a subgroup, or whatever.

Finally, it also feels better to separate concerns, and let the extension creator be responsible for error boundary and lazy loading, and perhaps something more in the future, but does not touch the specific way it is rendered.

/CC @iamEAP

Reply:
This has now been updated in 0cce460. The `createHomePageComponentExtension` extension creator has been updated to `createCardExtension` to be able to be more opinionated about how it will render (as a card). However, we still allow component developers to bypass this via the "Renderer"-prop, which allows integrators to render the component however they like. Two exaples of a tabbed layout as well as an accordion were provided in this same commit.